db/commitlog/commitlog.cc

```diff
@@ -1432,41 +1497,360 @@ future<R> db::commitlog::segment_manager::allocate_when_possible(T writer, db::t
+                        co_await s->close();
+                        continue;
+                    }
+
+                    auto rem2 = file_rem - (1 + file_rem/sector_size) * detail::sector_overhead_size;
+                    avail = std::min(rem2, max_mutation_size) 
+                        - segment::entry_overhead_size
+                        - segment::fragmented_entry_overhead_size
+                        - (pos == 0 ? segment::descriptor_header_size : 0)
+                        - segment::segment_overhead_size
+                        ;
+                }
+                if (!seg_ptr) {
+                    seg_ptr = s;
+                }
+
+                auto max_write = data_size - off;
+                auto to_write = std::min(avail, max_write);
+                auto rem = max_write - to_write;
+                partial_writer pw(writer, i, to_write, strm.read_view(to_write), rem, off, id);
+
+                switch (s->allocate(pw, fake_permit, timeout)) {
+                    case write_result::ok_need_batch_sync:
+                        s = co_await s->batch_cycle(timeout);
+                        [[fallthrough]];
+                    case write_result::ok:
+                        break;
+                    case write_result::must_sync:
+                        s = co_await with_timeout(timeout, s->sync());
+                        continue;
+                    case write_result::no_space:
+                        [[fallthrough]];
+                    case write_result::too_large:
+                        assert(0); // should not reach
+                        failed = true;
+                        break;
+                }
+
+                if (!failed) {
+                    clogger.debug("oversized wrote sub-entry {} fragment, id={} off={}, size={}, {} of {} bytes", i, id, off, to_write, off + to_write, data_size);
+
+                    off += to_write;
+                    if (s != seg_ptr) {
+                        // make first segment keep track of dependent data in
+                        // latter segments. Note, if we fail, all primary
+                        // rp_handles will expire, and we will free
+                        // the extended data as well. See release_cf_count.
+                        // Note also: In "normal" usage, we will not release
+                        // the references to other segments until the _whole_
+                        // segment is clean. This is intentional, because if 
+                        // something survives, the segment survives and might
+                        // be replayed, in which case we want to be able to
+                        // reconstruct a fragmented entry, if for no other
+                        // reason to be able to clear its state (see replay_state).
+                        seg_ptr->_extended_segments.emplace(pw._rp, std::move(pw._h));
+                    }
+                }
+            }
+        }
+    } catch (...) {
```

Comment:
Must we have such diverging code paths for single-fragment and multi-fragment entries?

Reply:
Yes. Because, as you might have missed, we need to lock out all other writes when doing this, otherwise we can't ensure being able to restore on failure etc, as well as be able to calculate whether we can even fit data (in a hard limit case).

Reply:
Maybe we can partially share code with helper functions.

Reply:
It does. There are code paths that look similar, but unfortunately they are subtly different, to such a degree that trying to share the code is most likely only making it less readable due to false abstraction/weird parameterizing/returns. Suggestions are of course welcome. 
(I am guessing the allocate-switch statements are the biggest eye sore. As you might recall, the code looks like this to ensure the fast path is as fast as possible in a coroutine universe...)

Reply:
Ok. I'll re-review. But why does it fail on CI?

Also, what's upgrade/downgrade story. I want the ability to downgrade even if there is an unclean shutdown (in fact the likely cause of a downgrade).

Reply:
Unless we backport the ability to read (but not produce) fragmented commit logs to earlier releases I don't see how that can work.

Reply:
Not sure about CI. The failure looks unrelated (failing to rm a non-empty dir). Building to run locally now.

Reply:
Cannot repro locally, I think the test failure is a python version error or something. shutil.rmtree is supposed (afaict) to remove any sub-folders etc - at least it does when I test it, but the test failed with the very error that it is not.